src/NUnitEngine/nunit.engine/ITestAgency.cs

```diff
@@ -34,5 +34,11 @@ public interface ITestAgency
         /// </summary>
         /// <param name="agent"></param>
         void Register(ITestAgent agent);
+
+        /// <summary>
+        /// Unregisters an agent with an agency
+        /// </summary>
+        /// <param name="agent"></param>
+        void Unregister(ITestAgent agent);
```

Comment:
Yes, it unfortunately still holding the reference to both the agent and runner which has me a little sad.  The comment in [TestAgency](https://github.com/nunit/nunit-console/blob/master/src/NUnitEngine/nunit.engine/Services/TestAgency.cs#L112) indicate there was intent for agents to be reused, but that doesn't appear to be happening in practice.

Given the current design, we could return a proxy ITestAgent that cleans up on Unregister.  But it might be better to revisit the design between ProcessRunner and TestAgency.  If ProcessRunner always just makes action calls to TestAgency, then we don't have this weird left over reference.

In other words,  ProcessRunner.RunTests => TestAgency.RunTests . TestAgency worries about either running the tests in a new process, or re-use existing.  But it just does actions.  It doesn't return agents like it does currently.